server/routes/volunteers.ts

```diff
@@ -144,6 +144,441 @@ router.route("/api/volunteers/").get(
+          url: lookupSlot,
+        });
+
+        if (slotFetch.kind === "error") {
+          res.status(INTERNAL_SERVER_ERROR).json({
+            message: slotFetch.error,
+          });
+
+          return;
+        }
+
+        if (slotFetch.records.length == 0) {
+          logger.error(
+            `Could not find a slot for ${supplierPickupEvent} for ${roleToAdd}`
+          );
+          res.status(BAD_REQUEST).json({
+            message: `Could not find a slot for ${supplierPickupEvent} for ${roleToAdd}`,
+          });
+
+          return;
+        }
+
+        const slotId = slotFetch.records[0].id;
+
+        slotPayload =
+          roleToAdd === "Driver"
+            ? { "Driving Slot": [slotId] }
+            : { "Logistics Slot": [slotId] };
+      } else if (
+        newRolesSet.has("Driver") &&
+        originalRolesSet.has("Distributor")
+      ) {
+        // Scenario 2: They switched from "Distributor" to "Driver".
+
+        // get the driver slot
+        const slotFetch = await airtableGET({
+          url: driverLookupSlot,
+        });
+
+        if (slotFetch.kind === "error") {
+          res.status(INTERNAL_SERVER_ERROR).json({
+            message: slotFetch.error,
+          });
+
+          return;
+        }
+
+        if (slotFetch.records.length == 0) {
+          logger.error(
+            `Could not find a driving slot for ${supplierPickupEvent}`
+          );
+          res.status(BAD_REQUEST).json({
+            message: `Could not find a driving slot for ${supplierPickupEvent}`,
+          });
+
+          return;
+        }
+
+        const slotId = slotFetch.records[0].id;
+
+        slotPayload = { "Driving Slot": [slotId], "Logistics Slot": null };
+      } else if (
+        newRolesSet.has("Distributor") &&
+        originalRolesSet.has("Driver")
+      ) {
+        // Scenario 2: They switched from "Driver" to "Distributor".
+
+        // get the logistics slot
+        const slotFetch = await airtableGET({
+          url: logisticsLookupSlot,
+        });
+
+        if (slotFetch.kind === "error") {
+          res.status(INTERNAL_SERVER_ERROR).json({
+            message: slotFetch.error,
+          });
+
+          return;
+        }
+
+        if (slotFetch.records.length == 0) {
+          logger.error(
+            `Could not find a logistics slot for ${supplierPickupEvent}`
+          );
+          res.status(BAD_REQUEST).json({
+            message: `Could not find a logistics slot for ${supplierPickupEvent}`,
+          });
+
+          return;
+        }
+
+        const slotId = slotFetch.records[0].id;
+
+        slotPayload = { "Driving Slot": null, "Logistics Slot": [slotId] };
+      } else {
+        // Scenario 4: They had one role and are keeping that same role.
+        // nothing needs to be done
+
+        res.status(OK).json({
+          message: `Volunteer ${volunteerId} has not changed their volunteer type.`,
+        });
+
+        return;
+      }
+    }
```

Comment:
The core logic here where you handle these cases is pretty long. Maybe we could simplify it a little bit by doing something like this? I haven't tested this out, so let me know if this doesn't work for some reason
```ts
    const originalRoles = [...originalInfo.records[0].fields.Type];
    const newRoles = [...airtableParticipantType];
    let slotPayload;
    if (originalRoles.length === 1 && newRoles.length === 2) {
      // 1. Volunteer had only one role and is now choosing both roles.
      let roleToAdd = originalRoles[0] === "Driver" ? "Distributor" : "Driver";

      // get matching slot, if they are becoming a driver we need to get them into the 10:30 slot,
      // packer only has one slot
      const lookupSlot =
        roleToAdd === "Driver" ? driverLookupSlot : logisticsLookupSlot;
      const slotFetch = await airtableGET({
        url: lookupSlot,
      });
      if (slotFetch.kind === "error") {
        res.status(INTERNAL_SERVER_ERROR).json({
          message: slotFetch.error,
        });
        return;
      }
      if (slotFetch.records.length === 0) {
        logger.error(
          `Could not find a slot for ${supplierPickupEvent} for ${roleToAdd}`
        );
        res.status(BAD_REQUEST).json({
          message: `Could not find a slot for ${supplierPickupEvent} for ${roleToAdd}`,
        });
        return;
      }

      const slotId = slotFetch.records[0].id;

      slotPayload =
        roleToAdd === "Driver"
          ? { "Driving Slot": [slotId] }
          : { "Logistics Slot": [slotId] };
    } else if (originalRoles.length === 2 && newRoles.length === 1) {
      // 5. Volunteer had both roles and is now choosing only one role.
      let roleToNullify = newRoles[0] === "Driver" ? "Distributor" : "Driver";
      slotPayload =
        roleToNullify === "Distributor"
          ? { "Driving Slot": null }
          : { "Logistics Slot": null };
    } else if (
      originalRoles.length === 1 &&
      newRoles.length === 1 &&
      originalRoles[0] !== newRoles[0]
    ) {
      // 2. Volunteer had only one role and is switching to the other single role.
      let newRole = originalRoles[0] === "Driver" ? "Distributor" : "Driver";
      // get the logistics slot
      const slotFetch = await airtableGET({
        url: newRole === "Driver" ? driverLookupSlot : logisticsLookupSlot,
      });

      if (slotFetch.kind === "error") {
        res.status(INTERNAL_SERVER_ERROR).json({
          message: slotFetch.error,
        });

        return;
      }

      if (slotFetch.records.length === 0) {
        logger.error(
          `Could not find a logistics slot for ${supplierPickupEvent}`
        );
        res.status(BAD_REQUEST).json({
          message: `Could not find a logistics slot for ${supplierPickupEvent}`,
        });

        return;
      }

      const slotId = slotFetch.records[0].id;

      slotPayload =
        newRole === "Driver"
          ? { "Driving Slot": [slotId], "Logistics Slot": null }
          : { "Driving Slot": null, "Logistics Slot": [slotId] };
    } else {
      //3 & 4
      logger.info(
        `Volunteer ${volunteerId} has not changed their volunteer type.`
      );
      res.status(OK).json({
        message: `Volunteer ${volunteerId} has not changed their volunteer type.`,
      });
      return;
    }
```

Reply:
This is really clean, I have implemented it in my new commit. Thank you so much for this suggestion.